### PR TITLE
tweak: Added more comprehensive mapped helper extensions to `SubstateDatabase` and its relatives

### PR DESCRIPTION
**Pull request by dhedey**
## Summary
* Added `SubstateDatabaseExtensions`, `CommittableSubstateDatabaseExtensions` and `ListableSubstateDatabaseExtensions` (in `radix-substate-store-interface/src/interface.rs`) which give easier interfaces to a substate store. Also replaced `MappedSubstateDatabase` etc with these.
* Removed `SpreadPrefixKeyMapper` where it could be removed in the codebase.

## Testing
Existing tests pass

## Update Recommendations
Make the following changes:
* `substate_database.get_mapped` to `substate_database.read_substate_typed`
* `substate_database.list_mapped` to `substate_database.read_entries_values_typed`.

### CODE REVIEW

File: radix-engine/src/track/track.rs [hunk 1]
```diff
@@ -62,7 +64,51 @@ pub struct TrackedSubstates {
     pub deleted_partitions: IndexSet<(NodeId, PartitionNumber)>,
 }
 
-impl<'s, S: SubstateDatabase, M: DatabaseKeyMapper + 'static> Track<'s, S, M> {
+impl TrackedSubstates {
+    pub fn to_state_updates(self) -> (IndexSet<NodeId>, StateUpdates) {
```
Comment (dhedey) [inline radix-engine/src/track/track.rs:68]: Moved from an importable function to a method

File: radix-substate-store-interface/src/interface.rs [hunk 2]
```diff
@@ -233,14 +267,304 @@ pub trait SubstateDatabase {
...
+        partition_number: PartitionNumber,
+        substate_key: impl ResolvableSubstateKey<'a>,
+    ) -> Option<T> {
+        let raw = self.read_substate(node_id, partition_number, substate_key)?;
+        Some(decode_value(&raw))
+    }
+
+    /// Use `from_substate_key_inclusive: None::<SubstateKey>` to read from start.
+    #[inline]
+    fn read_entries_unknown_key<'a>(
```
Comment (jakrawcz-rdx) [inline radix-substate-store-interface/src/interface.rs:343]: The `_unknown_key` sounds a bit weird.
Maybe align it with some other naming here? `_untyped_key`? or `_raw_key`?

File: radix-substate-store-interface/src/interface.rs [hunk 3]
```diff
@@ -233,14 +267,304 @@ pub trait SubstateDatabase {
...
+        from_substate_key_inclusive: impl ResolvableOptionalSubstateKey<'a>,
+    ) -> Box<dyn Iterator<Item = (DbSortKey, Vec<u8>)> + '_> {
+        self.list_entries_from(
+            &Self::db_partition_key(node_id, partition_number),
+            Self::optional_db_sort_key(from_substate_key_inclusive).as_ref(),
+        )
+    }
+
+    /// Use `from_substate_key_inclusive: None::<SubstateKey>` to read from start.
+    fn read_entries<'a, K: SubstateKeyContent>(
```
Comment (jakrawcz-rdx) [inline radix-substate-store-interface/src/interface.rs:356]: I thought we had a nice unwritten convention of `list_stuff()` when returning live (db-backed) iterator [and `read_stuff()` when fetching from db]. Is there any reason to abandon it?

**Review by jakrawcz-rdx — COMMENTED** [review]
The infra needed to handle this seems more bloated than expected. But I understand that it's because of multiple flavors of keys + multiple "reading modes" (wrt. mapping), so I can live with this 👍 

Were you able to sanity-test it against Node? (e.g. Engine State API's integration tests would cover a lot of that DB-reading, including corner cases that I previously caught and fixed)

_(I tried pulling this into Node, but there are too many non-obvious other changes on this branch, e.g. with intents)_

File: radix-substate-store-interface/src/interface.rs (hunk 3 above)
Comment (dhedey) [reply radix-substate-store-interface/src/interface.rs:356]: That's a good point! I'm keen to bikeshed names here.

Base names: `get_substate`, `list_entries`, `list_entries_from`
New names: ...?

The old abstractions were `get_mapped`. I was going for `read_` = new abstraction, but I don't really like it.

Maybe we just embrace `mapped`? And go for:
* `get_mapped_substate`
* `list_mapped_entries` (for `read_entries`) etc etc?

File: radix-substate-store-interface/src/interface.rs [hunk 4]
```diff
@@ -215,22 +240,375 @@ pub trait SubstateDatabase {
...
+/// These are a separate trait so that [`SubstateDatabase`] stays object-safe,
+/// and can be used as `dyn SubstateDatabase`.
+///
+/// Generic parameters aren't permitted on object-safe traits.
+pub trait SubstateDatabaseExtensions: SubstateDatabase {
+    /// Gets the raw bytes of the substate's value, if it exists.
+    ///
+    /// # Example
+    /// ```ignore
+    /// let is_bootstrapped = db.read_substate(
```
Comment (jakrawcz-rdx) [inline radix-substate-store-interface/src/interface.rs:266]: `get_raw_substate` I assume

File: radix-engine-tests/tests/application/stake_reconciliation.rs [hunk 5]
```diff
@@ -89,7 +89,7 @@ fn test_stake_reconciliation() {
     //     let partition_num = PartitionNumber(1);
     //     let substate_key = SubstateKey::Map(ScryptoRawPayload::from_valid_payload(vec![92, 32, 7, 32, 220, 0, 156, 5, 6, 83, 96, 189, 222, 100, 29, 145, 160, 147, 193, 127, 71, 54, 135, 62, 103, 35, 126, 168, 230, 117, 203, 71, 36, 132, 155, 157]));
     //     let value_from_database: ScryptoRawValue = ledger.substate_db()
-    //         .get_mapped::<SpreadPrefixKeyMapper, _>(&node_id, partition_num, &substate_key)
+    //         .read_substate_typed(&node_id, partition_num, substate_key)
```
Comment (jakrawcz-rdx) [inline radix-engine-tests/tests/application/stake_reconciliation.rs:92]: `get_raw_substate`?

File: radix-substate-store-interface/src/interface.rs [hunk 6]
```diff
@@ -215,22 +240,375 @@ pub trait SubstateDatabase {
...
+    // the specific kind of the substate key.
+    // ------------------------------------------------------------------------------------
+
+    /// Returns an iterator of the substates of a partition from an inclusive start cursor.
+    ///
+    /// The iterator returns `K` and the raw value for each substate.
+    /// The caller must specify `K` as [`FieldKey`], [`MapKey`] or [`SortedKey`].
+    ///
+    /// Pass `None::<SubstateKey>` as the cursor to iterate from the start of the partition.
+    fn list_kinded_raw_values<'a, K: SubstateKeyContent>(
```
Comment (jakrawcz-rdx) [inline radix-substate-store-interface/src/interface.rs:343]: the word "kinded" hurts my eyes, but I am not the native speaker here, and I also understand that you need some word to mean "with known kind of collection" (and I also don't have better ideas).

File: radix-substate-store-interface/src/interface.rs (hunk 4 above)
Comment (dhedey) [reply radix-substate-store-interface/src/interface.rs:266]: Thanks! Fixed in #1920

File: radix-engine-tests/tests/application/stake_reconciliation.rs (hunk 5 above)
Comment (dhedey) [reply radix-engine-tests/tests/application/stake_reconciliation.rs:92]: Thanks! Fixed in #1920
